[X86] Lower mathlib call ldexp into scalef when avx512 is enabled

KavinTheG opened the pull request:
Resolves #165694

KavinTheG committed:
[X86] Lower mathlib call ldexp into scalef when avx512 is enabled #165694

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -2590,6 +2590,26 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
+    
+    if (Subtarget.hasVLX()) {
+      for (MVT VT : { MVT::v4f32, MVT::v2f64, MVT::v8f32, MVT::v4f64 }) 
+        setOperationAction(ISD::FLDEXP, VT, Custom);
```
We might be better off allowing these on non-VLX targets and just widening (with zero) in the upper elements to 512-bits

RKSimon reviewed (changes requested):
Nice - a few minors

RKSimon commented on llvm/test/CodeGen/X86/ldexp-avx512.ll:
```diff
@@ -1,16 +1,21 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
 ; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512f | FileCheck %s --check-prefixes=CHECK,AVX512
-; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512vl | FileCheck %s --check-prefixes=CHECK,AVX512VL
+; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512vl,+avx512fp16 | FileCheck %s --check-prefixes=CHECK,AVX512VL
```
```
; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512f | FileCheck %s --check-prefixes=CHECK,AVX512,AVX512F
; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512f,+avx512fp16 | FileCheck %s --check-prefixes=CHECK,AVX512,AVX512FP16
; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512vl | FileCheck %s --check-prefixes=CHECK,AVX512VL,AVX512VLF
; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512vl,+avx512fp16 | FileCheck %s --check-prefixes=CHECK,AVX512VLFP16
```

KavinTheG committed:
Widen 128/256 bit vector types when AVX512VL is not available.

KavinTheG commented:
- Widened 128/256 bit vector types on non-VLX targets 
- Used SelectionDAG's helper abstractions (i.e., `ISD::INSERT_VECTOR_ELT` -> `getInsertVectorElt`)

KavinTheG committed:
Added fp16 vector handling to LowerFLDEXP.

KavinTheG commented:
- Added half vector handling, widening to f32 if fp16 is not available 
- Switched  to `widenSubVector`, removing the need for XVT and ExpVT
- Rebased to
  -  include the new tests for half vectors
  - squash previous clang format commits

RKSimon commented:
No, SCALEFS will only touch element[0] - not the upper elements of the 128-bit vector

KavinTheG committed:
Call SCALEF instead of SCALEFS for v4f32/v2f64 and refactored code.

KavinTheG committed:
Minor refactoring.

Switch getInsertVectorElts -> SCALAR_TO_VECTOR, and fixed styling issue.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -19153,6 +19155,81 @@ SDValue X86TargetLowering::LowerINSERT_VECTOR_ELT(SDValue Op,
+    if (Subtarget.hasFP16()) {
+      if (Subtarget.hasVLX()) {
+        Exp = DAG.getNode(ISD::SINT_TO_FP, DL, XTy, Exp);
+        return DAG.getNode(X86ISD::SCALEF, DL, XTy, X, Exp, X);
```
Why does SCALEF/S take 3 operands?

KavinTheG commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -19153,6 +19155,81 @@ SDValue X86TargetLowering::LowerINSERT_VECTOR_ELT(SDValue Op,
+  case MVT::v4f64:
+  case MVT::v16f32:
+  case MVT::v8f64:
+    Exp = DAG.getNode(ISD::SINT_TO_FP, DL, XTy, Exp);
```
After the switch, the only case where Exp is an integer is for vector fp16, but I do convert it to an fp after widening. I tried converting i16 vector exponents before widening, but that seemed very difficult. 

Would it be preferable to convert Exp  after the switch  statement like so? 

```
  case MVT::v8f64:
    // Exp = DAG.getNode(ISD::SINT_TO_FP, DL, XTy, Exp);
    if (XTy.getSizeInBits() == 512 || Subtarget.hasVLX()) {
      Exp = DAG.getNode(ISD::SINT_TO_FP, DL, XTy, Exp);
      return DAG.getNode(X86ISD::SCALEF, DL, XTy, X, Exp);
    }
    break;
    
 ...
 
  if (X.getValueType().getScalarType() != MVT::f16)
    Exp = DAG.getNode(ISD::SINT_TO_FP, DL, XTy, Exp);
  SDValue WideX = widenSubVector(X, true, Subtarget, DAG, DL, 512);
  SDValue WideExp = widenSubVector(Exp, true, Subtarget, DAG, DL, 512);

```

KavinTheG replied:
Alternatively for the vector fp16 cases, I realized it's possible to convert Exp to fp after extending it to i32. So Exp would be fp for all cases after the switch statement. 

```
    ExtVT = XTy.changeVectorElementType(MVT::f32);
    X = DAG.getFPExtendOrRound(X, DL, ExtVT);
    Exp = DAG.getSExtOrTrunc(Exp, DL, ExtVT.changeTypeToInteger());
    Exp = DAG.getNode(ISD::SINT_TO_FP, DL, ExtVT, Exp);
```

KavinTheG committed:
Ensure Exp is type integer after switch statement.

RKSimon reviewed (approved):
LGTM - @phoebewang any comments?

phoebewang reviewed (approved):
> LGTM - @phoebewang any comments?

LGTM.

KavinTheG committed:
Move FLDEXP setOperationAction above comment for clarity

KavinTheG committed:
Executed clang-format

RKSimon reviewed (changes requested):
Remove the empty line to fix the CI

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into x86_lower_ldexp
